src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py

```diff
@@ -514,46 +572,90 @@ def add_data_set(self, key=""):
         """
         A private class method to generate a new 1D or 2D data object based on
         the type of data within the set. Outside methods should call
         add_data_set() to be sure any existing data is stored properly.
 
         :param parent_list: List of names of parent elements
         """
-
-        if parent_list is None:
-            parent_list = []
-        if self._find_intermediate(parent_list, "Qx"):
+        if self._is2d(value):
             self.current_dataset = plottable_2D()
         else:
             x = np.array(0)
             y = np.array(0)
             self.current_dataset = plottable_1D(x, y)
         self.current_datainfo.filename = self.raw_data.filename
-
-    def _find_intermediate(self, parent_list, basename=""):
+        self.mask_name = ""
+        self.i_name = ""
+        self.i_node = ""
+        self.q_name = []
+        self.q_uncertainties = []
+        self.q_resolutions = []
+        self.i_uncertainties = ""
+
+    @staticmethod
+    def check_is_list_or_array(iterable):
+        try:
+            iter(iterable)
+            if (not isinstance(iterable, np.ndarray) and not isinstance(
+                    iterable, list)) or (isinstance(iterable, str) or
+                                         isinstance(iterable, unicode)):
+                raise TypeError
+        except TypeError:
+            iterable = iterable.split(",")
+        return iterable
+
+    def _find_data_attributes(self, value):
+        """
+        A class to find the indices for Q, the name of the Qdev and Idev, and
+        the name of the mask.
+        :param value: SASdata/NXdata HDF5 Group
+        """
+        attrs = value.attrs
+        signal = attrs.get("signal", "I")
+        i_axes = attrs.get("I_axes", ["Q"])
+        q_indices = attrs.get("Q_indices", [0])
+        q_indices = map(int, self.check_is_list_or_array(q_indices))
+        i_axes = self.check_is_list_or_array(i_axes)
+        keys = value.keys()
+        self.mask_name = attrs.get("mask")
+        for val in q_indices:
+            self.q_name.append(i_axes[val])
+        self.i_name = signal
+        self.i_node = value.get(self.i_name)
+        for item in self.q_name:
+            if item in keys:
+                q_vals = value.get(item)
+                if q_vals.attrs.get("uncertainties") is not None:
+                    self.q_uncertainties = q_vals.attrs.get("uncertainties")
+                elif q_vals.attrs.get("uncertainty") is not None:
+                    self.q_uncertainties = q_vals.attrs.get("uncertainty")
+                if isinstance(self.q_uncertainties, str):
+                    self.q_uncertainties = self.q_uncertainties.split(",")
+                if q_vals.attrs.get("resolutions") is not None:
+                    self.q_resolutions = q_vals.attrs.get("resolutions")
+                if isinstance(self.q_resolutions, (str, unicode)):
+                    self.q_resolutions = self.q_resolutions.split(",")
```

Comment:
Need a 2/3 compatible isstr function.

If these eventually want to be floats, may want to convert them now so that you can fail early in the loader.